tests/integration/security/authz/authz_test.go

```diff
@@ -0,0 +1,338 @@
+			runTest("[NestedToken1]", jwt.TokenIssuer1WithNestedClaims1, "/nested-key1", true)
+			runTest("[NestedToken2]", jwt.TokenIssuer1WithNestedClaims2, "/nested-key1", false)
+			runTest("[NestedToken1]", jwt.TokenIssuer1WithNestedClaims1, "/nested-key2", false)
+			runTest("[NestedToken2]", jwt.TokenIssuer1WithNestedClaims2, "/nested-key2", true)
+			runTest("[NestedToken1]", jwt.TokenIssuer1WithNestedClaims1, "/nested-2-key1", true)
+			runTest("[NestedToken2]", jwt.TokenIssuer1WithNestedClaims2, "/nested-2-key1", false)
+			runTest("[NestedToken1]", jwt.TokenIssuer1WithNestedClaims1, "/nested-non-exist", false)
+			runTest("[NestedToken2]", jwt.TokenIssuer1WithNestedClaims2, "/nested-non-exist", false)
+			runTest("[NoJWT]", "", "/tokenAny", false)
+
+			// Test condition "request.auth.principal" on path "/valid-jwt".
+			runTest("[NoJWT]", "", "/valid-jwt", false)
+			runTest("[Token1]", jwt.TokenIssuer1, "/valid-jwt", true)
+			runTest("[Token1WithAzp]", jwt.TokenIssuer1WithAzp, "/valid-jwt", true)
+			runTest("[Token1WithAud]", jwt.TokenIssuer1WithAud, "/valid-jwt", true)
+
+			// Test condition "request.auth.presenter" on suffix "/presenter".
+			runTest("[Token1]", jwt.TokenIssuer1, "/request/presenter", false)
+			runTest("[Token1WithAud]", jwt.TokenIssuer1, "/request/presenter", false)
+			runTest("[Token1WithAzp]", jwt.TokenIssuer1WithAzp, "/request/presenter-x", false)
+			runTest("[Token1WithAzp]", jwt.TokenIssuer1WithAzp, "/request/presenter", true)
+
+			// Test condition "request.auth.audiences" on suffix "/audiences".
+			runTest("[Token1]", jwt.TokenIssuer1, "/request/audiences", false)
+			runTest("[Token1WithAzp]", jwt.TokenIssuer1WithAzp, "/request/audiences", false)
+			runTest("[Token1WithAud]", jwt.TokenIssuer1WithAud, "/request/audiences-x", false)
+			runTest("[Token1WithAud]", jwt.TokenIssuer1WithAud, "/request/audiences", true)
+		})
+}
+
+func TestAuthz_WorkloadSelector(t *testing.T) {
+	framework.NewTest(t).
+		Features("security.authorization.workload-selector").
+		Run(func(t framework.TestContext) {
+			config.New().
+				// Allow /policy-<ns>-<svc> for all workloads.
+				// Also define a bad workload selector for path /policy-<ns>-<svc>-bad.
+				From(config.FileOrFail(t,
+					"testdata/v1beta1/workload.yaml.tmpl",
+					"testdata/v1beta1/workload-bad.yaml.tmpl")).
+				BuildToAll(t, nil, apps.Ns1AndNs2).
+
+				// Allow /policy-<ns>-all for all workloads.
+				From(config.FileOrFail(t, "testdata/v1beta1/workload-ns.yaml.tmpl")).
+				ToNamespace(apps.Ns1.Namespace).
+				Build(t).
+				From(config.FileOrFail(t, "testdata/v1beta1/workload-ns.yaml.tmpl")).
+				ToNamespace(apps.Ns2.Namespace).
+				Build(t).
+
+				// Allow /policy-istio-system-<svc> for all services in all namespaces. Just using ns1 to avoid
+				// creating duplicate resources.
+				From(config.FileOrFail(t, "testdata/v1beta1/workload-system-ns.yaml.tmpl")).
+				ToSystemNamespace(t).
```

Comment:
I cannot understand what this means. In the template they use `{{ .To.Config.Service }}`. How does that relate to SystemNamespace?

Reply:
Oh Its where its applied to. I feel like we might have gone a little too far on the chaining here - I was a big fan of the current state in master.

I don't see the benefits of this abstraction. It seems optimized to apply a bunch of config all at once in parallel, but I don't think that is important? Perhaps I am missing context on the motivation

Reply:
Yeah, some of the auth tests push configs that take a long time to complete. Really want to parallelize  it if possible.

Reply:
How is the `kubectl apply` slow? I would expect this to take milliseconds, its just an HTTP request?

Reply:
OK, so I dug in a bit and the problem with the security tests is that it ends up stacking a small delay (~ few seconds on GKE) across a lot of configs, so it can end up taking ~ 1 minute to push config. I've got a general fix here: https://github.com/istio/istio/pull/38024